pynobo.py

```diff
@@ -352,192 +476,188 @@ def send_command(self, commands):
-        bufsize -- maximum bytes to receive from the socket
-
-        Returns: a string list(s) of responses
-        """
-        response = b''
-        while response[-1:] != b'\r':
-            try:
-                response += self.client.recv(bufsize)
-            except socket.timeout:
-                # handshake needs to be sent every < 30 sec, preferably every 14 seconds
-                if keep_alive:
-                    now = time.time()
-                    if now - self.last_handshake > 14:
-                        self.send_command([self.API.HANDSHAKE])
-                        self.last_handshake = now
-                else:
-                    break
-            except ConnectionError as e:
-                self.logger.info('lost connection to hub (%s)', e)
-                self.socket_connected.clear()
-                break
-
-        # Handle more than one response in one receive
-        response_list = str(response, 'utf-8').split('\r')
-        response = []
-        for r in response_list:
-            if r:
-                response.append(r.split(' '))
-
-        return response
-
-    def socket_receive(self):
-        """The task running in daemon thread"""
-        while not self.socket_receive_exit_flag.is_set():
-            try:
-                if self.socket_connected.is_set():
-                    resp = self.get_response(True)
-                    for r in resp:
-                        self.logger.debug('received: %s', r)
+            _LOGGER.info('lost connection to hub (%s)', e)
+            await self.close()
 
-                        if r[0] == self.API.HANDSHAKE:
-                            pass # Handshake, no action needed
+    async def _get_initial_data(self):
+        self._received_all_info = False
+        await self.async_send_command([nobo.API.GET_ALL_INFO])
+        while not self._received_all_info:
+            self.response_handler(await self.get_response())
 
-                        elif r[0][0] == 'E':
-                            self.logger.error('error! what did you do? %s', r)
-                            #TODO: Raise something here?
-
-                        else:
-                            self.response_handler(r)
+    async def get_response(self):
+        """
+        Get a response string from the hub and reformat string list before returning it.
 
-                else:
-                    self.reconnect_hub()
-            except Exception as e:
-                # Ops, now we have real problems
-                self.logger.error("Unhandeled exception: %s", e, exc_info=1)
-                # Just disconnect (in stead of risking an infinite reconnect loop)
-                self.socket_receive_exit_flag.set()
+        :return: a single response as a list of strings where each string is a field
+        """
+        try:
+            message = await self._reader.readuntil(b'\r')
+            message = message[:-1]
+        except ConnectionError as e:
```

Comment:
I am also able to provoke TimeoutError at this point:
```
ERROR:pynobo:Unhandled exception [Errno 60] Operation timed out
<...>
TimeoutError: [Errno 60] Operation timed out
```
could we check for that as well?

Reply:
Yes, we can reconnect on timeout. But I'm curious to how this happened. Do you have steps to reproduce?

Reply:
I get this when disabling WIFI. After three to four `DEBUG:pynobo:sending: ['HANDSHAKE']` I get the exception, triggered from `self.get_response()`, line 513.

Reply:
Reproduced (also on macOS) by disabling ethernet after connection is made. Can also be reproduces by disconnecting ethernet cable, so this is a quite likely error to happen.

The timeout probably happens because asyncio sets a 60 seconds timeout on the reader, which is quite acceptable, since we expect a handshake every 14 seconds (recommended) to 30 seconds (max).

It was quite tricky to reconnect though, as the hub was rediscovered before local IP address was reassigned on DHCP, so I needed some error handling in `reconnect_hub` as well.

Reply:
The warning below is when the first call to `asyncio.open_connection` below failed with `ENETUNREACH`.
The `HELLO` message afterwards is after a successful call to `asyncio.open_connection`.

```
2021-04-28 21:25:04,738 - pynobo - DEBUG - sending: ['HANDSHAKE']
2021-04-28 21:25:18,742 - pynobo - DEBUG - sending: ['HANDSHAKE']
2021-04-28 21:25:32,743 - pynobo - DEBUG - sending: ['HANDSHAKE']
2021-04-28 21:25:46,745 - pynobo - DEBUG - sending: ['HANDSHAKE']
2021-04-28 21:25:58,830 - pynobo - INFO - Reconnecting due to [Errno 60] Operation timed out
2021-04-28 21:25:58,830 - pynobo - INFO - reconnecting to hub
2021-04-28 21:26:06,425 - pynobo - INFO - broadcast received: __NOBOHUB__102000013 from <ip redacted>
2021-04-28 21:26:07,843 - pynobo - WARNING - Failed to connect to ip set(): [Errno 51] Network is unreachable
2021-04-28 21:26:08,845 - pynobo - DEBUG - sending: ['HELLO', '1.1', <serial redacted>, '20210428212608']
```